helix-core/src/movement.rs

```diff
@@ -1,184 +1,254 @@
+                range.head.backwards_skip_newlines(slice)
+            };
+            let new_head = range.head.saturating_sub(1).start_of_block(slice);
+            Some(Range::new(new_anchor, new_head))
+        })?
+    };
+    (0..count).fold(range, |range, _| movement(range).unwrap_or(range))
+}
 
-        end = begin;
+pub fn move_next_word_end(slice: RopeSlice, range: Range, count: usize) -> Range {
+    let movement = |range: Range| -> Option<Range> {
+        let after_head = (range.head + 1).skip_newlines(slice);
+        (after_head + 1).inside(slice).then(|| {
+            let new_anchor = if range.head.is_boundary(slice) {
+                after_head
+            } else {
+                range.head.skip_newlines(slice)
+            };
+            let new_head = (range.head + 1).end_of_word(slice);
+            Some(Range::new(new_anchor, new_head))
+        })?
+    };
+    (0..count).fold(range, |range, _| movement(range).unwrap_or(range))
+}
 
-        with_end = skip_over_prev(slice, &mut end, char::is_whitespace);
+// ---- util ------------
 
-        // refetch
-        let ch = slice.char(end);
+/// Private helpers to help manipulate slice indices
+pub trait SliceIndexHelpers {
+    fn outside(&self, slice: RopeSlice) -> bool;
+    fn inside(&self, slice: RopeSlice) -> bool;
+    /// The next character after this belongs
+    /// to a different `Category`
+    fn is_boundary(&self, slice: RopeSlice) -> bool;
+    fn category(&self, slice: RopeSlice) -> Option<Category>;
+    /// Returns the start of a word/punctuation group followed by any amount of whitespace.
+    fn start_of_block(&self, slice: RopeSlice) -> Self;
+    /// Returns the end of a word/punctuation group followed by any amount of whitespace.
+    fn end_of_block(&self, slice: RopeSlice) -> Self;
+    /// Returns the end of a word/punctuation group.
+    fn end_of_word(&self, slice: RopeSlice) -> Self;
+    fn skip_newlines(&self, slice: RopeSlice) -> Self;
+    fn backwards_skip_newlines(&self, slice: RopeSlice) -> Self;
+    fn is_whitespace(&self, slice: RopeSlice) -> bool;
+    fn is_end_of_line(&self, slice: RopeSlice) -> bool;
+}
 
-        if is_word(ch) {
-            with_end = skip_over_prev(slice, &mut end, is_word);
-        } else if is_punctuation(ch) {
-            with_end = skip_over_prev(slice, &mut end, is_punctuation);
-        }
+impl SliceIndexHelpers for usize {
+    fn inside(&self, slice: RopeSlice) -> bool {
+        *self < slice.len_chars()
     }
 
-    Some(Range::new(begin, if with_end { end } else { end + 1 }))
-}
-
-pub fn move_next_word_end(slice: RopeSlice, mut begin: usize, count: usize) -> Option<Range> {
-    let mut end = begin;
+    fn outside(&self, slice: RopeSlice) -> bool {
+        !self.inside(slice)
+    }
 
-    for _ in 0..count {
-        if begin + 2 >= slice.len_chars() {
-            return None;
-        }
+    fn is_boundary(&self, slice: RopeSlice) -> bool {
+        (self + 1).inside(slice)
```

Comment:
Why do we need this check when it will be called?

Reply:
In the next line, we're calling `categorize` on `self+1`, so we need to ensure it's within bounds.

Reply:
But it is called outside the function already so we probably don't need to check next line.

Reply:
These redundant checks are trivially optimized away by the compiler, so it's generally better to have the extra safety. If someone in the future reuses `is_boundary` for something else in this module, we are averting a panic or a silent error.

Reply:
> These redundant checks are trivially optimized away by the compiler, so it's generally better to have the extra safety. If someone in the future reuses is_boundary for something else in this module, we are averting a panic or a silent error.

Are you sure about that?

Reply:
Of course!: https://godbolt.org/z/vMv5x79oP

The compiler eats this kind of redundancy for breakfast. Not just functions or modules, but even across crate boundaries, especially if you enable LTO. That's why I'm never an advocate for marking functions `inline` and other optimizer hints unless you're deep into optimization and looking at the results with a disassembler/profiler.

Reply:
Yeah, the simpler ones should be possible. I am just a bit afraid of the larger ones.

Reply:
https://godbolt.org/z/rYYPrzxs5 I think it still does compare

Reply:
It's an interesting point, and one that pertains to the approach to how the entire project is written.

In general, most developers' intuition of what the compiler can do falls *very, very short*. If you hold immutable references to data through an entire process, no matter how many functions, modules or even crates it spans, you have a near certainty that any redundancy will be optimized away. In fact, what's likely is that redundancies and inefficiencies you hadn't even considered will get optimized away (think things reordering, branch elimination, loop unrolling, etc).

My personal approach is usually not to worry about optmization until it's time to optimize, and when it's time to optimize to do it with proper benchmarking, looking at the generated assembly, identifying performance bottlenecks, etc. The risk of premature optimization over perceived, source-level inefficiencies is that it has a risk of unnecessarily complicating both the work and the future refactors, chasing performance problems that may not even be there.

It would be good for @archseer to establish a baseline for the project when it comes to performance and optimization, so we know what the expectation is.

Reply:
Yes, usually I don't worry about optimization but I worry about adding needless constructs that does a lot more than it need to. In this case which may potentially slows down performance. If we already always add those beforehand, we can just add a `debug_assert!` here just to double confirm, I don't think we need the full check.